src/lls_model.jl

```diff
@@ -98,23 +293,68 @@ function residual!(nls :: LLSModel, x :: AbstractVector, Fx :: AbstractVector)
   return Fx
 end
 
-function jac_structure_residual!(nls :: LLSModel, rows :: AbstractVector{<: Integer}, cols :: AbstractVector{<: Integer})
+function jac_structure_residual!(nls :: LLSMatrixModel, rows :: AbstractVector{<: Integer}, cols :: AbstractVector{<: Integer})
+  @lencheck nls.nls_meta.nnzj rows
+  @lencheck nls.nls_meta.nnzj cols
+  if issparse(nls.A)
+    I, J, V = findnz(nls.A)
+    rows .= I
+    cols .= J
+  else
+    m, n = size(nls.A)
+    I = ((i,j) for i = 1:m, j = 1:n)
+    rows .= getindex.(I, 1)[:]
+    cols .= getindex.(I, 2)[:]
+  end
+  return rows, cols
+end
+
+function jac_structure_residual!(nls :: LLSTripletModel, rows :: AbstractVector{<: Integer}, cols :: AbstractVector{<: Integer})
   @lencheck nls.nls_meta.nnzj rows
   @lencheck nls.nls_meta.nnzj cols
   rows .= nls.Arows
   cols .= nls.Acols
   return rows, cols
 end
 
-function jac_coord_residual!(nls :: LLSModel, x :: AbstractVector, vals :: AbstractVector)
+function jac_coord_residual!(nls :: LLSMatrixModel, x :: AbstractVector, vals :: AbstractVector)
+  @lencheck nls.meta.nvar x
+  @lencheck nls.nls_meta.nnzj vals
+  increment!(nls, :neval_jac_residual)
+  if issparse(nls.A)
+    vals .= nls.A.nzval
+  else
+    vals .= nls.A[:]
+  end
+  return vals
+end
+
+function jac_coord_residual!(nls :: LLSTripletModel, x :: AbstractVector, vals :: AbstractVector)
   @lencheck nls.meta.nvar x
   @lencheck nls.nls_meta.nnzj vals
   increment!(nls, :neval_jac_residual)
   vals .= nls.Avals
   return vals
 end
 
-function jprod_residual!(nls :: LLSModel, x :: AbstractVector, v :: AbstractVector, Jv :: AbstractVector)
+function jac_residual(nls :: LLSMatrixModel, x :: AbstractVector)
+  @lencheck nls.meta.nvar x
+  increment!(nls, :neval_jac_residual)
+  return nls.A
+end
+
+jac_op_residual(nls :: LLSOperatorModel, args...) = nls.A
+
+function jprod_residual!(nls :: AbstractLLSModel, x :: AbstractVector, v :: AbstractVector, Jv :: AbstractVector)
+  @lencheck nls.meta.nvar x
+  @lencheck nls.meta.nvar v
+  @lencheck nls.nls_meta.nequ Jv
+  increment!(nls, :neval_jprod_residual)
+  Jv .= nls.A * v
```

Comment:
So if A is an in-place operator, this copies the result?

Reply:
I think so.